### PR TITLE
merge w/ blanket branch updates

### PR DESCRIPTION
**Pull request by ElleryBaines**

### CODE REVIEW

File: plaidbench.py [hunk 1]
```diff
@@ -66,10 +60,14 @@ def elapsed(self):
 
 
 class Output(object):
-
     def __init__(self):
-        self.contents = None
-        self.precision = 'untested'
+      self.contents = None
```
Comment (brianretford) [inline plaidbench.py]: Indent is wrong here, should be +4

File: plaidbench.py [hunk 2]
```diff
@@ -80,6 +78,99 @@ def has_plaid():
...
+
+
+def run_intial(batch_size, compile_stop_watch, network, model):
+    print("Compiling and running initial batch, batch_size={}".format(batch_size))
+    compile_stop_watch.start()
+    optimizer = 'sgd'
+    if network[:3] == 'vgg':
+        from keras.optimizers import SGD
+        optimizer = SGD(lr=0.0001)
+    model.compile(optimizer=optimizer, loss='categorical_crossentropy',
```
Comment (brianretford) [inline plaidbench.py]: It's confusing, but 'compile' in keras doesn't actually compile tile code. If you read through the old code, you'll see that the compile stop watch measures one call to train and one call to infer before stopping the compile watch and starting the main one. It's impossible with the current way things are setup to isolate compilation time, so currently compile measures compilation + 1 batch of inference or training. You'll have to remove this function and just have the infer and train functions, and pass the compile stop watch to both.

File: plaidbench.py [hunk 3]
```diff
@@ -109,12 +114,20 @@ def train(x_train, y_train, epoch_size, model, batch_size, compile_stop_watch, e
...
+    compile_stop_watch.start_outer()
+    stop_watch.start_outer()
+    
+    run_intial(batch_size, compile_stop_watch, network, model);
     y = model.predict(x=x_train, batch_size=batch_size)
     
     compile_stop_watch.stop()
+    
     output.contents = y
     printf('Warmup')
```
Comment (brianretford) [inline plaidbench.py]: I actually think you can remove the 'warmup' part. (lines 132-135)

File: plaidbench.py [hunk 4]
```diff
@@ -127,6 +140,8 @@ def inference(network, model, batch_size, compile_stop_watch, output, x_train, e
         stop_watch.stop()
         time.sleep(.025 * random.random())
 
+    stop_watch.stop()
```
Comment (brianretford) [inline plaidbench.py]: This is extraneous

File: plaidbench.py [hunk 5]
```diff
@@ -80,6 +78,111 @@ def has_plaid():
         return False
 
 
+def value_check(examples, epochs, batch_size):
+    if epochs > examples:
+        raise ValueError('The number of epochs must be less than the number of examples.')
```
Comment (earhart) [inline plaidbench.py]: What if they're equal?  Either use >=, or change the message.

File: plaidbench.py [hunk 6]
```diff
@@ -80,6 +78,111 @@ def has_plaid():
...
+    if batch_size > (examples // epochs):
+        raise ValueError('The number of examples per epoch must be greater than the batch size.')
```
Comment (earhart) [inline plaidbench.py]: Again, what if they're equal?  Either the condition or the message is busted.

File: plaidbench.py [hunk 7]
```diff
@@ -80,6 +78,111 @@ def has_plaid():
...
+    if examples % epochs != 0:
+        raise ValueError('The number of examples must be divisible by the number of epochs.')
+    if (examples // epochs) % batch_size != 0:
+        raise ValueError('The number of examples per epoch is not divisble by the batch size.')
```
Comment (earhart) [inline plaidbench.py]: "divisible"
And maybe, "must be divisible by the batch size"?

File: plaidbench.py [hunk 8]
```diff
@@ -80,6 +78,111 @@ def has_plaid():
...
+
+
+def train(x_train, y_train, epoch_size, model, batch_size, compile_stop_watch, 
+    epochs, stop_watch, output, network):
```
Comment (earhart) [inline plaidbench.py]: Either align these vertically (so put 'epochs' under 'x_train'), or use a hanging indent with four spaces and _no_ parameters on the first line.  Here's the [style guide rule](https://google.github.io/styleguide/pyguide.html?showone=Indentation#Indentation).

File: plaidbench.py [hunk 9]
```diff
@@ -80,6 +78,111 @@ def has_plaid():
...
+    # Training
+    compile_stop_watch.start_outer()
+    stop_watch.start_outer()
+    
+    run_intial(batch_size, compile_stop_watch, network, model)
```
Comment (earhart) [inline plaidbench.py]: "run_initial"

File: plaidbench.py [hunk 10]
```diff
@@ -80,6 +78,111 @@ def has_plaid():
...
+    stop_watch.stop()
+
+
+def inference(network, model, batch_size, compile_stop_watch, output, x_train, 
+    examples, stop_watch):
+    # Inference
+    compile_stop_watch.start_outer()
+    stop_watch.start_outer()
+    
+    run_intial(batch_size, compile_stop_watch, network, model);
```
Comment (earhart) [inline plaidbench.py]: "run_initial"

File: plaidbench.py [hunk 11]
```diff
@@ -80,6 +78,111 @@ def has_plaid():
...
+    module = os.path.join(this_dir, 'networks', '%s.py' % module)
+    globals = {}
+    exec_(open(module).read(), globals)
+    x_train = globals['scale_dataset'](x_train)
+    model = globals['build_model']()
+    printf("Model loaded.")
+    return module, x_train, model
+
+
+def run_intial(batch_size, compile_stop_watch, network, model):
```
Comment (earhart) [inline plaidbench.py]: "run_initial"

File: plaidbench.py [hunk 12]
```diff
@@ -80,6 +78,111 @@ def has_plaid():
...
+
+
+def train(x_train, y_train, epoch_size, model, batch_size, compile_stop_watch, 
+    epochs, stop_watch, output, network):
+    # Training
+    compile_stop_watch.start_outer()
+    stop_watch.start_outer()
+    
+    run_intial(batch_size, compile_stop_watch, network, model)
+    x = x_train[:epoch_size]
```
Comment (earhart) [inline plaidbench.py]: What are you doing with x and y here?  Move them later, to make it more obvious what they're for.

File: plaidbench.py [hunk 13]
```diff
@@ -80,6 +78,111 @@ def has_plaid():
...
+    y = y_train[:epoch_size]
+    model.train_on_batch(x_train[0:batch_size], y_train[0:batch_size])
+    
+    compile_stop_watch.stop()
+
+    for i in range(epochs):
+        if i == 1:
```
Comment (earhart) [inline plaidbench.py]: I'm confused -- what does "Doing the main timing" mean?  It doesn't look like this changes anything about when you start and stop the stopwatch?

File: plaidbench.py [hunk 14]
```diff
@@ -80,6 +78,111 @@ def has_plaid():
...
+            printf('Doing the main timing')
+        stop_watch.start()
+        history = model.fit(
+            x=x, y=y, batch_size=batch_size, epochs=1, shuffle=False, initial_epoch=0)
+        stop_watch.stop()
+        time.sleep(.025 * random.random())
+        if i == 0:
+            output.contents = [history.history['loss']]
+    output.contents = np.array(output.contents)
```
Comment (earhart) [inline plaidbench.py]: So, are you intentionally only capturing the first loss?  It might be nice to add a comment about that; it looks wrong.

File: plaidbench.py [hunk 15]
```diff
@@ -80,6 +78,111 @@ def has_plaid():
...
+
+    stop_watch.stop()
+
+
+def inference(network, model, batch_size, compile_stop_watch, output, x_train,
```
Comment (earhart) [inline plaidbench.py]: [Bad arg formatting](https://google.github.io/styleguide/pyguide.html?showone=Indentation#Indentation)

File: plaidbench.py [hunk 16]
```diff
@@ -80,6 +78,111 @@ def has_plaid():
...
+    stop_watch.start_outer()
+    
+    run_intial(batch_size, compile_stop_watch, network, model);
+    y = model.predict(x=x_train, batch_size=batch_size)
+    
+    compile_stop_watch.stop()
+    
+    output.contents = y
+
+    for i in range(32 // batch_size + 1):
```
Comment (earhart) [inline plaidbench.py]: What is '32' here?  (Add a comment?)

File: plaidbench.py [hunk 17]
```diff
@@ -111,125 +216,131 @@ def main():
...
+
+        # Run network
+        try:
+            # Setup
+            x_train, y_train = setup(args.train, epoch_size, batch_size)
+
+            # Loading the model
+            module, x_train, model = load_model(args.module, x_train)
+
+            # training run
```
Comment (earhart) [inline plaidbench.py]: Maybe move this comment into the if: block

File: plaidbench.py [hunk 18]
```diff
@@ -111,125 +216,131 @@ def main():
...
+            if args.train:
+                value_check(examples, epochs, batch_size)
+                train(x_train, y_train, epoch_size, model, batch_size, 
+                    compile_stop_watch, epochs, stop_watch, output, network)
+            # inference run
```
Comment (earhart) [inline plaidbench.py]: Move this one under the else:

File: plaidbench.py [hunk 19]
```diff
@@ -111,125 +216,131 @@ def main():
...
+            x_train, y_train = setup(args.train, epoch_size, batch_size)
+
+            # Loading the model
+            module, x_train, model = load_model(args.module, x_train)
+
+            # training run
+            if args.train:
+                value_check(examples, epochs, batch_size)
+                train(x_train, y_train, epoch_size, model, batch_size, 
+                    compile_stop_watch, epochs, stop_watch, output, network)
```
Comment (earhart) [inline plaidbench.py]: [Fix indentation](https://google.github.io/styleguide/pyguide.html?showone=Indentation#Indentation)

File: plaidbench.py [hunk 20]
```diff
@@ -111,125 +216,131 @@ def main():
...
+            # Record data
+            if args.blanket_run:
+                network_data['average_example_duration'] = execution_duration / examples
+
+            network_data['execution_duration'] = execution_duration
+            network_data['compile_duration'] = compile_duration
+            network_data['precision'] = output.precision
+
+            # Print statement
+            printf('Example finished, elapsed: {} (compile), {} (execution)'.format(
```
Comment (earhart) [inline plaidbench.py]: In general, it's probably better to use keyword arguments over positional for format strings.

File: plaidbench.py [hunk 21]
```diff
@@ -111,125 +216,131 @@ def main():
...
+            module, x_train, model = load_model(args.module, x_train)
+
+            # training run
+            if args.train:
+                value_check(examples, epochs, batch_size)
+                train(x_train, y_train, epoch_size, model, batch_size, 
+                    compile_stop_watch, epochs, stop_watch, output, network)
+            # inference run
+            else:
+                inference(args.module, model, batch_size, compile_stop_watch,
```
Comment (flaub) [inline plaidbench.py]: It would be nice to do some argument validation for inference as well. One check would be that the `examples` is divisible by `batch_size`.

File: plaidbench.py [hunk 22]
```diff
@@ -80,6 +78,113 @@ def has_plaid():
...
+
+def train(x_train, y_train, epoch_size, model, batch_size, compile_stop_watch, 
+          epochs, stop_watch, output, network):
+    # Training
+    compile_stop_watch.start_outer()
+    stop_watch.start_outer()
+    
+    run_initial(batch_size, compile_stop_watch, network, model)
+    model.train_on_batch(x_train[0:batch_size], y_train[0:batch_size])
+    x = x_train[:epoch_size]
```
Comment (earhart) [inline plaidbench.py]: Move the `x` and `y` initializers after the `compiler_stop_watch.stop()` call.
